---
id: components
title: Views Types
slug: /components
---
import ComponentImage from '../src/pages/_ComponentImage';

Vitessce supports several __view types__ for visualization of single-cell data, and several view types for updating visualization parameters. The view names found on this page are all valid values for the field [`layout[].component`](/docs/view-config-json/#layout) in the view config. On this page you will also find screenshots of each view.

### `scatterplot`

The scatterplot view displays 2-dimensional (pre-computed) embeddings / projections (such as t-SNE or UMAP).
Use with the `embeddingType` coordination type to specify which embedding to use when mapping cells onto scatterplot points.
Use with datasets containing the `cells` data type. Optionally, the `cell-sets` and `genes` data types can be used for coloring cells on the plot.

<ComponentImage filename="scatterplot-2.png" alt="Screenshot of Scatterplot View" />

### `heatmap`

The heatmap view displays a (normalized) cell-by-gene (or gene-by-cell) matrix visualization.
This view uses the `expression-matrix` data type, and optionally displays cell set color assignments if a file with the `cell-sets` data type is available in the same dataset.
The heatmap can support very large matrices (~6,000 x ~9,000) when using Zarr-based file types (best performance can be acheived when arrays are saved with the type `uint8`).

<ComponentImage filename="heatmap.png" alt="Screenshot of Heatmap View" />
<ComponentImage filename="heatmap-2.png" alt="Screenshot of Heatmap View" />

### `spatial`

The spatial view is meant to display data with spatial coordinates, including spatially-resolved cell segmentations as polygons (from the `cells` data type) and molecule positions as points (from the `molecules` data type). The spatial view also includes a multiplexed and multi-scale image viewer (implemented with [Viv](http://viv.gehlenborglab.org/)). Image files can be defined with the `raster` data type. Cell segmentations can be colored by gene expression data through the `expression-matrix` data type or cell set color assignments through the `cell-sets` data type.

<ComponentImage filename="spatial-2.png" alt="Screenshot of Spatial View" />
<ComponentImage filename="spatial.png" alt="Screenshot of Spatial View" />

### `layerController`

The layer controller view provides an interface for manipulating the visualization layers displayed in the `spatial` view.
The image layers are coordinated through the [`spatialRasterLayers`](/docs/coordination-types/#spatialRasterLayers) coordination type.

<ComponentImage filename="layerController.png" alt="Screenshot of Layer Controller View" />

### `genomicProfiles`

The genomic profiles view displays genome browser tracks (using the `genomic-profiles` data type) containing bar plots, where the genome is along the x-axis and the value at each genome position is encoded with a bar along the y-axis.
Genome tracks may be colored by corresponding cell set color assignments if the `cell-sets` data type is available in the same dataset, coordinated on the [`cellSetSelection`](/docs/coordination-types/#cellSetSelection) coordination type.
This view was implemented with [HiGlass](https://higlass.io/).

<ComponentImage filename="genomicProfiles.png" alt="Screenshot of Genomic Profiles View" />

### `genes`

The genes view displays an interactive list of genes when used with datasets containing the `expression-matrix` data type.

<ComponentImage filename="genes.png" alt="Screenshot of Expression Levels View" />

### `cellSets`

The cell sets view displays an interactive list of (potentially hierarchical) cell sets when used with datasets containing the `cell-sets` data type. This can be useful for managing cell sets representing clustering algorithm outputs or cell type annotations.

<ComponentImage filename="cellSets.png" alt="Screenshot of Cell Set Manager View" />

### `cellSetSizes`

The cell set sizes view displays a bar plot with the currently-selected cell sets (using the `cell-sets` data type and [`cellSetSelection`](/docs/coordination-types/#cellSetSelection) coordination type) on the x-axis and bars representing their size (by number of cells) on the y-axis. This view was implemented with [Vega-Lite](https://vega.github.io/vega-lite/).

<ComponentImage filename="cellSetSizes.png" alt="Screenshot of Cell Set Sizes View" />

### `description`

The description view can be used to display text content. This may be useful for communicating the details of a dataset or a data analysis process. When the `spatial` view is used for visualization of imaging data, the `description` view also renders a dropdown containing image metadata.

<ComponentImage filename="description.png" alt="Screenshot of Description View" />

### `status`

The status view displays debugging messages, including app-wide error messages when datasets fail to load or when schemas fail to validate. Details about the entity under the mouse cursor (cell, gene, and/or molecule) are displayed during hover interactions.

<ComponentImage filename="status.png" alt="Screenshot of Status View" />

### `cellSetExpression`

The cell set expression view displays a violin plot with expression values for the selected gene, implemented with [Vega-Lite](https://vega.github.io/vega-lite/).

<ComponentImage filename="cellSetExpression.png" alt="Screenshot of Cell Set Expression View" />

### `expressionHistogram`

The expression histogram view displays a bar plot with the distribution of expression values for the selected gene, implemented with [Vega-Lite](https://vega.github.io/vega-lite/).

<ComponentImage filename="expressionHistogram.png" alt="Screenshot of Expression Histogram View" />
