diff --git a/lib/index.js b/lib/index.js
index c823b80af766d09f33b6ee505484a93cd2715c5d..7a0f81d833932c7e3a8f268ec69604b9d8315cb7 100644
--- a/lib/index.js
+++ b/lib/index.js
@@ -584,11 +584,12 @@ var Neuroglancer = /*#__PURE__*/function (_React$Component) {
           onViewerStateChanged = _this$props2.onViewerStateChanged,
           callbacks = _this$props2.callbacks,
           ngServer = _this$props2.ngServer,
-          key = _this$props2.key;
+          key = _this$props2.key,
+          bundleRoot = _this$props2.bundleRoot;
       this.viewer = (0, _neuroglancer.setupDefaultViewer)({
         brainMapsClientId: brainMapsClientId,
         target: this.ngContainer.current,
-        bundleRoot: "/"
+        bundleRoot: bundleRoot || "/"
       });
       this.setCallbacks(callbacks);
 
