diff --git a/dist/src/index.js b/dist/src/index.js
index a450563be06a52877aea180c966e823a5c579426..5c30935342c682abfeb6adafc49008ed612b2133 100644
--- a/dist/src/index.js
+++ b/dist/src/index.js
@@ -1,4 +1,4 @@
 export { default as FetchStore } from "./fetch.js";
-export { default as FileSystemStore } from "./fs.js";
+//export { default as FileSystemStore } from "./fs.js";
 export { default as ReferenceStore } from "./ref.js";
 export { default as ZipFileStore } from "./zip.js";
