---
id: components
title: Views Types
slug: /components
---
import ComponentImage from '../src/pages/_ComponentImage';

Vitessce supports several __view types__ for visualization of single-cell data, and several view types for updating visualization parameters. The view names found on this page are all valid values for the field [`layout[].component`](/docs/view-config-json/#layout) in the view config. On this page you will also find screenshots of each view.

### `scatterplot`

The scatterplot view displays 2-dimensional (pre-computed) embeddings / projections (such as t-SNE or UMAP).
Use with the `embeddingType` coordination type to specify which embedding to use when mapping cells onto scatterplot points.
Use with datasets containing the `obsEmbedding` data type. Optionally, the `obsSets` and `obsFeatureMatrix` data types can be used for coloring cells on the plot.

<ComponentImage filename="scatterplot-2.png" alt="Screenshot of Scatterplot View" />

### `gating`
The gating view displays 2-dimensional scatterplot of gene expression data.
Users select two genes, and a scatterplot is dynamically generated using the `obsFeatureMatrix` data. Gating can then be performed by using the lasso or box select tools.
Use with datasets containing the `obsFeatureMatrix` data types. Optionally, the `obsSets` and `obsFeatureMatrix` data types can be used for coloring cells on the plot.

<ComponentImage filename="gating-1.png" alt="Screenshot of Gating View" />
<ComponentImage filename="gating-2.png" alt="Screenshot of Gating View" />

### `heatmap`

The heatmap view displays a (normalized) cell-by-gene (or gene-by-cell) matrix visualization.
This view uses the `obsFeatureMatrix` data type, and optionally displays cell set color assignments if a file with the `obsSets` data type is available in the same dataset.
The heatmap can support very large matrices (~6,000 x ~9,000) when using Zarr-based file types (best performance can be acheived when arrays are saved with the type `uint8`).

<ComponentImage filename="heatmap.png" alt="Screenshot of Heatmap View" />
<ComponentImage filename="heatmap-2.png" alt="Screenshot of Heatmap View" />

### `spatial`

The spatial view is meant to display data with spatial coordinates, including spatially-resolved cell segmentations as polygons (from the `obsSegmentations` data type) and molecule positions as points (from the `obsLocations` data type). The spatial view also includes a multiplexed and multi-scale image viewer (implemented with [Viv](http://viv.gehlenborglab.org/)). Image files can be defined with the `image` data type. Cell segmentations can be colored by gene expression data through the `obsFeatureMatrix` data type or cell set color assignments through the `obsSets` data type.

<ComponentImage filename="spatial-2.png" alt="Screenshot of Spatial View" />
<ComponentImage filename="spatial.png" alt="Screenshot of Spatial View" />

### `layerController`

The layer controller view provides an interface for manipulating the visualization layers displayed in the `spatial` view.
The image layers are coordinated through the [`spatialImageLayer`](/docs/coordination-types/#spatialImageLayer) coordination type.

<ComponentImage filename="layerController.png" alt="Screenshot of Layer Controller View" />

### `genomicProfiles`

The genomic profiles view displays genome browser tracks (using the `genomic-profiles` data type) containing bar plots, where the genome is along the x-axis and the value at each genome position is encoded with a bar along the y-axis.
Genome tracks may be colored by corresponding cell set color assignments if the `obsSets` data type is available in the same dataset, coordinated on the [`obsSetSelection`](/docs/coordination-types/#obsSetSelection) coordination type.
This view was implemented with [HiGlass](https://higlass.io/).

<ComponentImage filename="genomicProfiles.png" alt="Screenshot of Genomic Profiles View" />

### `featureList`

The feature list view displays an interactive list of features (i.e., genes when `featureType: 'gene'`).
The list is obtained from the feature axis of matrix files corresponding to the `obsFeatureMatrix` data type.

<ComponentImage filename="genes.png" alt="Screenshot of Expression Levels View" />

### `obsSets`

The observation sets view displays an interactive list of (potentially hierarchical) observation sets (e.g., cell sets) when used with datasets containing the `obsSets` data type. This can be useful for managing cell sets representing clustering algorithm outputs or cell type annotations.

<ComponentImage filename="cellSets.png" alt="Screenshot of Cell Set Manager View" />

### `obsSetSizes`

The observation set sizes view displays a bar plot with the currently-selected observation sets (e.g., cell sets) using the `obsSets` data type and [`obsSetSelection`](/docs/coordination-types/#obsSetSelection) coordination type on the x-axis and bars representing their size (e.g., by number of cells) on the y-axis. This view was implemented with [Vega-Lite](https://vega.github.io/vega-lite/).

<ComponentImage filename="cellSetSizes.png" alt="Screenshot of Cell Set Sizes View" />

### `description`

The description view can be used to display text content. This may be useful for communicating the details of a dataset or a data analysis process. When the `spatial` view is used for visualization of imaging data, the `description` view also renders a dropdown containing image metadata.

<ComponentImage filename="description.png" alt="Screenshot of Description View" />

### `status`

The status view displays debugging messages, including app-wide error messages when datasets fail to load or when schemas fail to validate. Details about the entity under the mouse cursor (cell, gene, and/or molecule) are displayed during hover interactions.

<ComponentImage filename="status.png" alt="Screenshot of Status View" />

### `obsSetFeatureValueDistribution`

The observation set feature value distribution view displays a violin plot with values (e.g., expression values) for the selected feature (e.g., gene), implemented with [Vega-Lite](https://vega.github.io/vega-lite/).

<ComponentImage filename="cellSetExpression.png" alt="Screenshot of Cell Set Expression View" />

### `featureValueHistogram`

The feature value histogram view displays the distribution of values (e.g., expression) for the selected feature (e.g., gene), implemented with [Vega-Lite](https://vega.github.io/vega-lite/).

<ComponentImage filename="expressionHistogram.png" alt="Screenshot of Expression Histogram View" />

### `Neuroglancer`

The Neuroglancer view embeds [Neuroglancer](https://github.com/google/neuroglancer) - a high-performance, WebGL2-based viewer developed by Google for huge multi-resolution volumes and segmentations.

<ComponentImage filename="neuroglancer.png" alt="Screenshot of Neuroglancer view" />

The view requires a minimal viewerState configuration with the following properties that can be added to the Vitessce config using the `setProps({ viewerState: VIEWER_STATE})` call.
A complete schema of the `viewerState` can be found [here](https://neuroglancer-docs.web.app/json/api/index.html).

```
const VIEWER_STATE = {
  dimensions: { x: [1, "nm"], y: [1, "nm"], z: [1, "nm"] }, // Global coordinate system. Values are [scale, unit].

  // Camera setting
  position: [0, 0, 0],                 // position [x, y, z] in dimension units
  projectionOrientation: [0, 0, 0, 1], // rotation in xyz space in quaternion units
  projectionScale: 12000,              // zoom level

  layout: "3d", // a 3d layout is used for Vitessce integration
  // Layers â€” include at least one that actually draws something.
  layers: [
    {
      name: "image_name",
      type: "image",
      source: "precomputed://<IMAGE_SOURCE>", // e.g. a precomputed pyramid
    },
    // Segmentation (only shows if segments are specified here or dynamically via `obsSets`)
    {
        name: "seg_name",
        type: "segmentation",
        source: "precomputed://<SEG_SOURCE>",
        segments: ["1","2"], // optional; can be empty array
        segmentColors : {
            "1": "#ff00ff",
            "2": "#0000ff"
        }
    }
  ]
};

```